File: src/main/java/io/github/bot/terminal/application/domains/bot_commands/worker/BotCommandsWorker.java
package io.github.bot.terminal.application.domains.bot_commands.worker;


import io.github.bot.terminal.application.domains.bot_commands.entity.BotCommand;
import io.github.bot.terminal.application.domains.bot_commands.factory.BotCommandsFactory;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Service;

import java.util.concurrent.ExecutorService;

@Slf4j
@Service
@RequiredArgsConstructor
class BotCommandsWorker {

    private final BotCommandsFactory factory;
    private final ExecutorService executorService;

    @Scheduled(fixedDelay = 7_000)
    public void botCommandsScheduler() {
        for (BotCommand botCommand : factory.getAll()) {
            if (!botCommand.isEnabled()) {
                continue;
            }
            executorService.execute(() -> botCommand.proceed());
        }
    }

}

File: src/main/kotlin/io/github/support/terminal/application/domains/workers/BotCommandsWorker.kt
package io.github.support.terminal.application.domains.workers

import io.github.support.terminal.application.domains.bot_commands.entity.BotCommand
import io.github.support.terminal.application.domains.bot_commands.entity.BotCommandsFactory
import io.github.support.terminal.application.domains.notificarion_api.entity.NotificationApi
import io.github.support.terminal.application.domains.notificarion_api.entity.NotificationApiFactory
import org.slf4j.LoggerFactory
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Service
import java.lang.invoke.MethodHandles
import java.util.*
import java.util.concurrent.Callable
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

@Service
class BotCommandsWorker(
        private val notificationApiFactory: NotificationApiFactory,
        private val botCommandsFactory: BotCommandsFactory,
        private val executorService: ExecutorService
) {

    @Scheduled(fixedDelay = 7000)
    fun scheduler() {
        val callables: MutableList<Callable<Any>> = ArrayList()
        for (notificationApi in notificationApiFactory.all()) {
            callables.add(
                    Executors.callable { handle(notificationApi) }
            )
        }
        executorService.invokeAll(callables)
    }

    private fun handle(notificationApi: NotificationApi) {

        val cmdMap: Map<String, List<BotCommand>> = botCommandsFactory.byNotificationApiId(notificationApi.id)
                .groupBy { it.cmd() }

        for (message in notificationApi.lastMessages()) {
            val cmd = message.text.split(" ")[0].trim()
            if(cmd.isEmpty()) {
                continue
            }
            cmdMap[cmd]?.let {
                it.forEach { botCommand: BotCommand ->
                    log.info("Start execution of command: `${cmd}` (${message.text})")
                    botCommand.performAction(message.text).notifyAboutResult(notificationApi)
                }
            }
        }
    }

    companion object {
        private val log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass())
    }
}